### PR TITLE
Add a new contiguous allocator strategy option

### PR DESCRIPTION
**Pull request by smarterclayton**
Also fix an error in port allocator.  I'm using contiguous allocation for a UID
and SELinux label allocator in OpenShift but will be backporting it post 1.0.

@justinsb something I missed during your code review or @thockin

### CODE REVIEW

File: pkg/registry/service/portallocator/allocator.go [hunk 1]
```diff
@@ -151,8 +151,7 @@ func (r *PortAllocator) Restore(pr util.PortRange, data []byte) error {
 	if !ok {
 		return fmt.Errorf("not a snapshottable allocator")
 	}
-	snapshottable.Restore(pr.String(), data)
-	return nil
+	return snapshottable.Restore(pr.String(), data)
```
Comment (justinsb) [inline pkg/registry/service/portallocator/allocator.go]: Oh wow - I think you caught this one before as well, but I must have lost it during the git operations.  I will do a diff to see what else I may have lost....

File: pkg/registry/service/allocator/bitmap.go [hunk 2]
```diff
@@ -66,6 +67,30 @@ func NewAllocationMap(max int, rangeSpec string) *AllocationBitmap {
...
+		count:     0,
+		max:       max,
+		rangeSpec: rangeSpec,
+	}
+	return &a
+}
+
+// NewRandomAllocationInterface creates an allocation bitmap satisfying Interface using the
+// random scan strategy.
+func NewRandomAllocationInterface(max int, rangeSpec string) Interface {
```
Comment (justinsb) [inline pkg/registry/service/allocator/bitmap.go]: Do we need these two ...Interface functions?

**Comment by justinsb** [comment]
I'm very wary of reusing IPs or ports too quickly... my fear is that pod2 comes up on the same IP or port as pod1 was on, and sees traffic meant for pod1.

Admittedly randomizing just makes the problem less likely, but this strategy will expose it every time if it is a genuine problem.  As for whether that is a good thing or a bad thing, I don't know :-)

No objection to the code, though I don't really get where it is going.

And thank you for catching the dropped err retval (again!)

File: pkg/registry/service/portallocator/allocator.go (hunk 1 above)
Comment (justinsb) [inline pkg/registry/service/portallocator/allocator.go]: Ah - this is a different function than the others, so not a lost commit, just an error I made repeatedly in this file and failed to check for other instances of.  Thanks for catching!

File: pkg/registry/service/allocator/bitmap.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/registry/service/allocator/bitmap.go]: These methods implement allocator.Factory directly so you don't have to wrap the method.  Convenience.

**Comment by smarterclayton** [comment]
This doesn't change existing call sites, it just adds the new option.  For security related allocations which are much larger than 10k (average service subnet) it limits the amount of contiguous storage needed.  For instance, the selinux allocator with two labels is 535k possible values.  

I would rather expose overlap issues early for security related items.  That's much more dangerous.

**Comment by erictune** [comment]
@smarterclayton do you hope to get this into 1.0?

**Comment by smarterclayton** [comment]
Yes, it introduces no changes to running function, and makes reuse easier.  It should be a simple merge.

**Comment by justinsb** [comment]
OK, thanks for clarifying.  LGTM, though the PR where we switch this on will have some interesting security discussions!

**Comment by smarterclayton** [comment]
Here's how the contiguous allocator is being used today: https://github.com/openshift/origin/pull/2544/files#diff-3999d64b0b8fa9423b47f9feb61db2b0R1007

File: pkg/registry/service/allocator/bitmap.go [hunk 3]
```diff
@@ -166,3 +191,17 @@ func randomScanStrategy(allocated *big.Int, max, count int) (int, bool) {
...
 	return 0, false
 }
+
+// contiguousScanStrategy tries to allocate starting at 0 and filling in any gaps
+func contiguousScanStrategy(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
+		return 0, false
+	}
+	for i := 0; i < max; i++ {
+		at := i % max
```
Comment (thockin) [inline pkg/registry/service/allocator/bitmap.go]: Why do you need this mod, if your loop ends at max anyway?

**Comment by thockin** [comment]
LGTM, just a nit I don't understand.

File: pkg/registry/service/allocator/bitmap.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/registry/service/allocator/bitmap.go]: Copy and paste error.  Fixed

----- Original Message -----

**Comment by thockin** [comment]
LGTM

**Pull request merged**